### PR TITLE
Sebastian's solution

### PR DESCRIPTION
**Pull request by sebastianquek**
## Setup
* Ensure you have node on your machine (v14 or later)
* Run the following:
```bash
npm i
npm test
```

## Notes
- Relies on TypeScript, Jest and Cheerio
- Large changeset is mostly because of the creation of `package-lock.json`
- Main files to look at are `src/scrape-google-carousel.ts` and `src/__tests__/scrape-google-carousel.test.ts`
- 2 additional test files are added (`files/leonardo-da-vinci-art.html` and `files/top-cyclists.html`)

## Possible future improvements
Given the time constraints, there are a couple of things that can possibly be explored.
- Better test the links and images to avoid flakey tests when a new html is generated
- Figure out a better way to decode the image URIs, i.e. `\x3d` -> `=`
- Perform requests to actual search URLs
- Look into more variants of Google Search carousels (e.g. is it possible to have multiple extensions?)
- `null` images when Google's placeholder image is used

### CODE REVIEW

File: src/scrape-google-carousel.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+   *
+   * Preconditions:
+   * - JavaScript is defined after the definition of the carousel
+   * - Within JavaScript, image URI (i.e. data:image...) is defined before the image ID
+   * - base64 image sources
+   *
+   * @param element
+   * @returns
+   */
+  private getImage(element: cheerio.Element): string | null {
```
Comment (aciddjus) [inline src/scrape-google-carousel.ts:91]: What about extracting images that are hidden on the initial page load:

image

Your current implementation returns `null`

image

Would you be able to update your code to extract those as well?

**Review by aciddjus — CHANGES_REQUESTED** [review]
Thank you, Sebastian. 💯 

Realy interesting solution using TypeScrip. And great job on adding tests and new HTMLs.

I just have a few remarks:

----------------------------

Currently, the tests are failing:

image

image

File: src/scrape-google-carousel.ts (hunk 1 above)
Comment (sebastianquek) [reply src/scrape-google-carousel.ts:88]: Thanks for your patience, just got back from vacation and have worked on the changes.

image

The image URIs are extracted via the target anchor element's id.

**Comment by sebastianquek** [comment]
While I was debugging the image issue, I noticed a couple of regressions on the playground,

1. Name of first artwork isn't appearing (i.e. "Mona Lisa")
2. Multi-line names are not appearing correctly (i.e. "The Last Supper")

image

https://serpapi.com/searches/2a4ab588f281c6f4/6306e8aafe41d0b5d036d61c.json

Should I raise an issue here: https://github.com/serpapi/public-roadmap/issues?

**Comment by hartator** [comment]
> Should I raise an issue here: https://github.com/serpapi/public-roadmap/issues?

Yes, that will be great. We will address these. Thank you @sebastianquek. 👍

File: src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  }
+  return [
+    tokens[0],
+    ...tokens
+      .slice(1)
+      .map(
+        (h) => String.fromCharCode(parseInt(h.slice(0, 2), 16)) + h.slice(2)
+      ),
+  ].join("");
+};
```
Comment (ilyazub) [inline src/utils.ts:14]: :+1:

Usually this is handled with `str.replace(/* ... */)`, but you approach looks fine.

File: src/scrape-google-carousel.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+   * Preconditions:
+   * - image URI is set via the `data-src` attribute
+   *
+   * @param elementId
+   * @returns
+   */
+  private extractImageUriUsingElementId(elementId: string): string | null {
+    const targetInnerHtml = this.html.match(`'${elementId}','(.*?)'`)?.[1];
+    if (targetInnerHtml) {
+      const parsedInnerHtml = cheerio.load(hexToAscii(targetInnerHtml));
```
Comment (ilyazub) [inline src/scrape-google-carousel.ts:156]: Good job on decoding and unescaping HTML.

**Review by ilyazub — COMMENTED** [review]
:+1:

**Pull request closed without merging**